[BUG] Sink Worker Threads encounter fatal exception on unknown type

**Describe the bug**
A misdefined parquet schema lead to an exception that shutdown the sink threads
```
2023-08-21T18:35:22.570 [log-pipeline-sink-worker-2-thread-2] ERROR org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor - Pipeline [log-pipeline] process worker encountered a fatal exception, cannot proceed further
java.util.concurrent.ExecutionException: org.apache.avro.SchemaParseException: "s3" is not a defined name. The type of the "s3" field must be a defined name or a {"type": ...} expression.
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
	at java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[?:?]
	at org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor.afterExecute(PipelineThreadPoolExecutor.java:70) ~[data-prepper-core-2.4.0-SNAPSHOT.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1129) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: org.apache.avro.SchemaParseException: "s3" is not a defined name. The type of the "s3" field must be a defined name or a {"type": ...} expression.
	at org.apache.avro.Schema.parse(Schema.java:1716) ~[avro-1.11.1.jar:1.11.1]
	at org.apache.avro.Schema.parse(Schema.java:1805) ~[avro-1.11.1.jar:1.11.1]
	at org.apache.avro.Schema$Parser.parse(Schema.java:1469) ~[avro-1.11.1.jar:1.11.1]
	at org.apache.avro.Schema$Parser.parse(Schema.java:1457) ~[avro-1.11.1.jar:1.11.1]
	at org.opensearch.dataprepper.plugins.codec.parquet.ParquetOutputCodec.parseSchema(ParquetOutputCodec.java:142) ~[s3-sink-2.4.0-SNAPSHOT.jar:?]
	at org.opensearch.dataprepper.plugins.codec.parquet.ParquetOutputCodec.buildSchemaAndKey(ParquetOutputCodec.java:81) ~[s3-sink-2.4.0-SNAPSHOT.jar:?]
	at org.opensearch.dataprepper.plugins.codec.parquet.ParquetOutputCodec.start(ParquetOutputCodec.java:69) ~[s3-sink-2.4.0-SNAPSHOT.jar:?]
	at org.opensearch.dataprepper.plugins.sink.s3.S3SinkService.output(S3SinkService.java:113) ~[s3-sink-2.4.0-SNAPSHOT.jar:?]
	at org.opensearch.dataprepper.plugins.sink.s3.S3Sink.doOutput(S3Sink.java:109) ~[s3-sink-2.4.0-SNAPSHOT.jar:?]
	at org.opensearch.dataprepper.model.sink.AbstractSink.lambda$output$0(AbstractSink.java:64) ~[data-prepper-api-2.4.0-SNAPSHOT.jar:?]
	at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:141) ~[micrometer-core-1.10.5.jar:1.10.5]
	at org.opensearch.dataprepper.model.sink.AbstractSink.output(AbstractSink.java:64) ~[data-prepper-api-2.4.0-SNAPSHOT.jar:?]
	at org.opensearch.dataprepper.pipeline.Pipeline.lambda$publishToSinks$5(Pipeline.java:336) ~[data-prepper-core-2.4.0-SNAPSHOT.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	... 2 more
```

**To Reproduce**
S3 sink config
```
codec:
          parquet:
            schema: >
              [{
                "type" : "record",
                "namespace" : "org.opensearch.dataprepper.examples",
                "name" : "S3",
                "fields" : [
                  { "name" : "bucket", "type" : "string"},
                  { "name" : "key", "type": "string"}
                ]
              },
              {
                "type" : "record",
                "namespace" : "org.opensearch.dataprepper.examples",
                "name" : "Parquet",
                "fields" : [
                  ...
                  { "name" : "s3", "type": "s3"}, # This was the problem, the type should be S3
                  { "name" : "@timestamp", "type" : "string"}
                ]
              }]

```

**Expected behavior**
An error/warning should be logged or a metric emitted rather than the thread encountering a fatal exception

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Environment (please complete the following information):**
 - OS: [e.g. Ubuntu 20.04 LTS]
 - Version [e.g. 22]

**Additional context**
Add any other context about the problem here.

--- a/data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/s3/S3SinkService.java
+++ b/data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/s3/S3SinkService.java
@@ -112,28 +112,28 @@ public class S3SinkService {
         try {
             for (Record<Event> record : records) {
 
-                if (currentBuffer.getEventCount() == 0) {
-                    final Event eventForSchemaAutoGenerate = record.getData();
-                    codec.start(currentBuffer.getOutputStream(), eventForSchemaAutoGenerate, codecContext);
-                }
-
                 final Event event = record.getData();
-                codec.writeEvent(event, currentBuffer.getOutputStream());
-                int count = currentBuffer.getEventCount() + 1;
-                currentBuffer.setEventCount(count);
+                try {
+                    if (currentBuffer.getEventCount() == 0) {
+                        codec.start(currentBuffer.getOutputStream(), event, codecContext);
+                    }
+
+                    codec.writeEvent(event, currentBuffer.getOutputStream());
+                    int count = currentBuffer.getEventCount() + 1;
+                    currentBuffer.setEventCount(count);
+                } catch (Exception ex) {
+                    LOG.error("Unable to add event to buffer. Dropping this event.");
+                }
 
                 if (event.getEventHandle() != null) {
                     bufferedEventHandles.add(event.getEventHandle());
                 }
                 flushToS3IfNeeded();
             }
-        } catch (IOException e) {
-            LOG.error("Exception while write event into buffer :", e);
+            flushToS3IfNeeded();
+        } finally {
+            reentrantLock.unlock();
         }
-
-        flushToS3IfNeeded();
-
-        reentrantLock.unlock();
     }
 
     private void releaseEventHandles(final boolean result) {

--- a/data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/s3/S3SinkService.java
+++ b/data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/s3/S3SinkService.java
@@ -121,13 +121,17 @@ public class S3SinkService {
                     codec.writeEvent(event, currentBuffer.getOutputStream());
                     int count = currentBuffer.getEventCount() + 1;
                     currentBuffer.setEventCount(count);
+
+                    if (event.getEventHandle() != null) {
+                        bufferedEventHandles.add(event.getEventHandle());
+                    }
                 } catch (Exception ex) {
+                    if (event.getEventHandle() != null) {
+                        event.getEventHandle().release(false);
+                    }
                     LOG.error("Unable to add event to buffer. Dropping this event.");
                 }
 
-                if (event.getEventHandle() != null) {
-                    bufferedEventHandles.add(event.getEventHandle());
-                }
                 flushToS3IfNeeded();
             }
             flushToS3IfNeeded();

--- a/data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/s3/S3SinkServiceTest.java
+++ b/data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/s3/S3SinkServiceTest.java
@@ -485,7 +485,6 @@ class S3SinkServiceTest {
         final OutputStream outputStream = mock(OutputStream.class);
         when(buffer.getOutputStream()).thenReturn(outputStream);
 
-        final S3SinkService s3SinkService = createObjectUnderTest();
 
         List<Record<Event>> records = generateEventRecords(2);
         Event event1 = records.get(0).getData();
@@ -493,12 +492,17 @@ class S3SinkServiceTest {
 
         doThrow(RuntimeException.class).when(codec).writeEvent(event1, outputStream);
 
-        s3SinkService.output(records);
+        createObjectUnderTest().output(records);
 
         InOrder inOrder = inOrder(codec);
         inOrder.verify(codec).start(eq(outputStream), eq(event1), any());
         inOrder.verify(codec).writeEvent(event1, outputStream);
         inOrder.verify(codec).writeEvent(event2, outputStream);
+
+        verify(event1.getEventHandle()).release(false);
+        verify(event1.getEventHandle(), never()).release(true);
+        verify(event2.getEventHandle()).release(true);
+        verify(event2.getEventHandle(), never()).release(false);
     }
 
     @Test

--- a/data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/s3/S3SinkServiceTest.java
+++ b/data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/s3/S3SinkServiceTest.java
@@ -10,6 +10,7 @@ import io.micrometer.core.instrument.DistributionSummary;
 import org.apache.commons.lang3.RandomStringUtils;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.mockito.InOrder;
 import org.opensearch.dataprepper.metrics.PluginMetrics;
 import org.opensearch.dataprepper.model.codec.OutputCodec;
 import org.opensearch.dataprepper.model.configuration.PluginModel;
@@ -54,6 +55,7 @@ import static org.mockito.ArgumentMatchers.anyLong;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.doNothing;
 import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.inOrder;
 import static org.mockito.Mockito.lenient;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.never;
@@ -469,6 +471,34 @@ class S3SinkServiceTest {
         for (EventHandle eventHandle : eventHandles2) {
             verify(eventHandle).release(true);
         }
+    }
+
+    @Test
+    void output_will_skip_and_drop_failed_records() throws IOException {
+        bufferFactory = mock(BufferFactory.class);
+        final Buffer buffer = mock(Buffer.class);
+        when(bufferFactory.getBuffer(any(S3Client.class), any(), any())).thenReturn(buffer);
+
+        final long objectSize = random.nextInt(1_000_000) + 10_000;
+        when(buffer.getSize()).thenReturn(objectSize);
+
+        final OutputStream outputStream = mock(OutputStream.class);
+        when(buffer.getOutputStream()).thenReturn(outputStream);
+
+        final S3SinkService s3SinkService = createObjectUnderTest();
+
+        List<Record<Event>> records = generateEventRecords(2);
+        Event event1 = records.get(0).getData();
+        Event event2 = records.get(1).getData();
+
+        doThrow(RuntimeException.class).when(codec).writeEvent(event1, outputStream);
+
+        s3SinkService.output(records);
+
+        InOrder inOrder = inOrder(codec);
+        inOrder.verify(codec).start(eq(outputStream), eq(event1), any());
+        inOrder.verify(codec).writeEvent(event1, outputStream);
+        inOrder.verify(codec).writeEvent(event2, outputStream);
     }
 
     @Test

